### PR TITLE
Added bus-analytics scheduler process

### PR DESCRIPTION
**Pull request by MattBuckley-TfN**
# Changes

Added scheduler script to connect to the database get the required inputs and run the [otp4gb-py](https://github.com/Transport-for-the-North/otp4gb-py) package to produce bus travel times. The script will insert the data back into the database.

# Tasks

- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Has documentation (including user guide) been updated
- [ ] Have new dependencies been added (or changed) in relevant `requirements.txt`
- [ ] Code linting, tests and other workflows pass

### CODE REVIEW

**Comment by MattBuckley-TfN** [comment]
@hulsiejames please could you have a light touch review so we can get this merged into main. We've both looked at it in the past but if you've got time to have another glance and make sure there's nothing I've missed.

File: scheduler.py [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+    "num_nans_transfers",
+    "median_generalised_cost",
+    "min_generalised_cost",
+    "max_generalised_cost",
+    "num_nans_generalised_cost",
+    "run_id",
+    "timetable_id",
+    "ID",
+    "zone_type_id",
+]
```
Comment (hulsiejames) [inline scheduler.py:243]: Is this meant to be here? doesn't seem to be used

File: scheduler.py [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+        main()
+    except Exception as exc:
+        print(f"Ignoring exception: {exc}")
+
+    # Check if shutdown has been updated since program started
+    try:
+        shutdown = _get_env_bool(_SHUTDOWN_ENV_VARIABLE, _SHUTDOWN)
+    except ValueError as exc:
+        print(f"Reverting to previous value ({_SHUTDOWN}) - {exc}")
+        shutdown = _SHUTDOWN
```
Comment (hulsiejames) [inline scheduler.py:439]: I guess this gives the user time to realise it defaults to shutdown on completion, so they can change the env variable whilst the process is running?

But we don't seem to explicitly let the user know that it will shutdown by default as this isn't stated in main, or passed to any process within main, unless I have missed something!

File: scheduler.py [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+    folder: pathlib.Path = config.ASSET_DIR,
+) -> TimetableData:
+    """Get GTFS timetable from database and save to `folder`."""
+    timetable = pg_database.find_recent_timetable(adjusted)
+    if timetable is None:
+        raise ValueError("couldn't find a timetable")
+
+    path = shutil.copy(timetable.actual_timetable_path, folder)
+
+    date = next_day(timetable.upload_date, 0)
```
Comment (hulsiejames) [inline scheduler.py:132]: Is there a possibility that we manually upload a historic timetable and the upload_date is out of range of the timetable?

Don't think this is an issue though as I think the timetable downloader should upload the timetable itself so would only be an issue if there was an old timetable manually uploaded which I don't think will happen much if at all.

File: scheduler.py [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+    if timetable is None:
+        raise ValueError("couldn't find a timetable")
+
+    dst_path = folder / timetable.actual_timetable_path.name
+    if dst_path.is_file():
+        dst_path.unlink()
+
+    shutil.copy(timetable.actual_timetable_path, dst_path)
+
+    date = next_day(timetable.upload_date, 0)
```
Comment (hulsiejames) [inline scheduler.py:156]: Is there a possibility that we manually upload a historic timetable and the upload_date is out of range of the timetable?

Don't think this is an issue though as I think the timetable downloader should upload the timetable itself so would only be an issue if there was an old timetable manually uploaded which I don't think will happen much if at all

**Review by hulsiejames — APPROVED** [review]
Looks good! 

Couple of comments out of interest.

File: scheduler.py (hunk 1 above)
Comment (MattBuckley-TfN) [reply scheduler.py:243]: Nice spot, getting rid of it

File: scheduler.py (hunk 4 above)
Comment (MattBuckley-TfN) [reply scheduler.py:156]: Switching to using the timetables update time.

File: scheduler.py (hunk 2 above)
Comment (MattBuckley-TfN) [reply scheduler.py:439]: Added message when tool starts

**Comment by MattBuckley-TfN** [comment]
@hulsiejames thanks for your review I've addressed those comments and merged
